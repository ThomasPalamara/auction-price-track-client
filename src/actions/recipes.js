export const selectRecipe = (recipe = {}) => (
  {
    type: 'SELECT_RECIPE',
    recipe,
  });


const defaultRecipes = [{
  professions: [''],
  type: '',
  craft: {
    blizzardId: 0,
    quantity: 1,
  },
  reagents: [{
    blizzardId: 0,
    quantity: 0,
  }, {
    blizzardId: 1,
    quantity: 0,
  }],
}];

export const addRecipes = (recipes = defaultRecipes) => (
  {
    type: 'ADD_RECIPES',
    recipes,
  }
);
