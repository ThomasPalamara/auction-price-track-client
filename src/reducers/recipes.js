
export default (state = [], action) => {
    switch (action.type) {
        case 'ADD_RECIPES':
            return action.recipes;
        case 'SELECT_RECIPE':
            return {...state, selectedRecipe: action.recipe};
        default:
            return state;
    }
}

