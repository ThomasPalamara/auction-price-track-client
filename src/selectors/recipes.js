export default (list, { text, profession, isCustom}) => {
	console.log(list);
	return list.filter((item) => {
		// const textMatch = item.description.toLowerCase().includes(text.toLowerCase());
        const professionMatch = profession !== 'all' ? item.professions.indexOf(profession) != -1 : true ;
        const isCustomMatch = item.isCustom ? item.isCustom === isCustom : false === isCustom;
		return isCustomMatch && professionMatch;
	});
};
