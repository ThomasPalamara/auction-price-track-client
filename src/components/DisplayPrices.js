import React from "react";
import { connect } from 'react-redux';
import DisplayPricesTable from './DisplayPricesTable';
import DisplayPricesTotal from './DisplayPricesTotal';
import { Row, Col } from 'antd';
import { editQuantityIngredient } from '../actions/ingredient';
import { editQuantityProduct } from '../actions/product';


const DisplayPrices = (props) => (
    <div>
        <Row>
            <Col span={24}>
                <DisplayPricesTotal auctionHouse={props.auctionHouse} />
            </Col>
        </Row>
        <Row>
            <Col span={12}>
                <DisplayPricesTable auctionHouse={props.auctionHouse} editQuantity={editQuantityProduct} items={props.product} />
            </Col>
            <Col span={12}>
                <DisplayPricesTable auctionHouse={props.auctionHouse} editQuantity={editQuantityIngredient} items={props.ingredient} />
            </Col>
        </Row>
    </div>
);

const mapStatetoProps = state => {

    return {
        product: state.product,
        ingredient: state.ingredient

    }
};

export default connect(mapStatetoProps)(DisplayPrices);


//Tableau de toutes les auctions pour chaque item
// {this.props.product &&
//     this.props.product.map((item) => {
//         const itemElem = this.props.auctionHouse.find(elem => elem.item_id === item.id.toString())
//         const totalPrice = item.quantity
//         return (
//             <div>
//                 <p>{item.name}</p>
//                 <InputNumber min={0} defaultValue={1}
//                     onChange={(value) => {
//                         this.props.dispatch(editQuantityProduct({ id: item.id, quantity: value }))
//                     }} />
//             </div>
//         )
//     })}