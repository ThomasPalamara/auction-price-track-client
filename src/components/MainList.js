import React from "react";
import { connect } from 'react-redux';
import MainListItem from "./MainListItem";
import MainListFilter from "./MainListFilter";
import FlipMove from "react-flip-move";
import { Row, Col, Spin, Icon } from "antd";
import ItemLanguage from "./ItemLanguage";
import selectRecipes from "../selectors/recipes"
import { addRecipes } from "../actions/recipes"

//reagentJson = reagentJson.reagents; Voir avec LUDO

const antIcon = <Icon type="loading" style={{ fontSize: 24 }} spin />;

export class MainList extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      itemsJSON: null,
      recipesJSON: null,
      loading: true,
    }
  }

  componentDidUpdate() {
    window.$WowheadPower.refreshLinks();
  }
  componentWillMount() {
    fetch(`/api/recipes`)
      .then(res => res.json())
      .then(response => {
        this.setState({ loading: false });
        this.props.dispatch(addRecipes(response));
      });
  }
  componentDidMount() {
    fetch(`/api/items`)
      .then(res => res.json())
      .then(response => {
        this.setState({ response: response });
      });
  }


  render() {
    return (
      <div className="MainList">
        <h2 className="step">2/ Choose your Recipe</h2>

        <MainListFilter activeList={this.state.activeList} handleToggleList={this.handleToggleList} />

        <ItemLanguage />

        <Row className="container-item-display">
          <Col span={24}>
            <Spin spinning={this.state.loading} indicator={antIcon}>
              <ul className="list">
                <FlipMove>
                  {this.props.recipesList &&
                    this.props.recipesList
                      .map((item, i) => {
                        return (<MainListItem element={item} key={item._id} itemLanguage={this.props.itemLanguage} />)
                      })
                  }
                </FlipMove>
              </ul>

            </Spin>
          </Col>
        </Row>

      </div>
    )
  }
};

const mapStatetoProps = state => {
  return {
    itemLanguage: state.itemLanguage,
    recipesList: selectRecipes(state.recipes, state.filters)
    // recipesList : state.recipes
  }
};

//	More common way to do is to export by default for the HOC and call the function as parameters
export default connect(mapStatetoProps)(MainList);