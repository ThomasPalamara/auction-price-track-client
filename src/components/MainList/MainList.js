import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import FlipMove from 'react-flip-move';
import { Row, Col, Spin, Icon, Divider } from 'antd';
import MainListItem from './MainListItem';
import MainListFilter from './MainListFilter';
import ItemLanguage from '../ItemLanguage';
import selectRecipes from '../../selectors/recipes';
import { addRecipes } from '../../actions/recipes';
import { apiURL, recipePropTypes } from '../../constants';

const antIcon = <Icon type="loading" style={{ fontSize: 24 }} spin />;

// Cannot use hook without breaking the WH link refresher. Inside a hook they are updating infinitly.
// Maybe the update need to be done inside the mainlistitem component to avoid that.
// But it also mean calling the function for each and every items instead of calling it one time here

export class MainList extends React.Component {
  constructor(props) {
    super(props);
    this.state = { loading: true };
  }

  // Adding all recipe to the store.
  componentWillMount() {
    const { dispatch } = this.props;

    fetch(`${apiURL}/recipes`)
      .then((res) => {
        if (!res.ok) {
          throw new Error('Not found');
        }
        return res.json();
      })
      .then((response) => {
        this.setState({ loading: false });
        dispatch(addRecipes(response));
      });
  }

  // Function provided by WowHead allowing us to refresh the icons and tooltip provided by them.
  // We need to update everytime a recipe appear or disappear due to filters.
  componentDidUpdate() {
    window.$WowheadPower.refreshLinks();
  }

  render() {
    const { recipesList, itemLanguage, selectedRecipe } = this.props;
    const { loading } = this.state;
    return (
      <div className="mainList">
        <Divider orientation="left">
          <h3 className="step">2/ Choose your Items</h3>
        </Divider>

        <MainListFilter />

        <Row className="mainList__container">
          <Col span={24}>
            <Spin spinning={loading} indicator={antIcon}>
              <ul className="mainList__list">
                <FlipMove>
                  {recipesList
                    && recipesList
                      .map(item => (
                        <MainListItem
                          element={item}
                          key={item._id}
                          itemLanguage={itemLanguage}
                          active={(selectedRecipe && selectedRecipe._id === item._id) || false}
                        />
                      ))
                  }
                </FlipMove>
              </ul>
            </Spin>
          </Col>
        </Row>
      </div>
    );
  }
}

MainList.propTypes = {
  recipesList: PropTypes.arrayOf(PropTypes.object).isRequired,

  itemLanguage: PropTypes.string,

  selectedRecipe: recipePropTypes,
  dispatch: PropTypes.func.isRequired,
};

MainList.defaultProps = { itemLanguage: 'en', selectedRecipe: null };

const mapStatetoProps = state => ({
  itemLanguage: state.itemLanguage,
  recipesList: selectRecipes(state.recipes.recipesList, state.filters),
  selectedRecipe: state.recipes.selectedRecipe,
});

// More common way to do is to export by default for the HOC and call the function as parameters
export default connect(mapStatetoProps)(MainList);
