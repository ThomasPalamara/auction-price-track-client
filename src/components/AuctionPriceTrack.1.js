import React, { useState } from 'react';
import ReactSwipe from 'react-swipe';

import RealmSelection from './RealmSelection';
import MainList from './MainList';
import Stats from './Stats';

// TODO Ajouter la selection de language au store puis corrigÃ© celle en place et ajouter a display table

const AuctionPriceTrack = () => {
  // State par default pour faciliter les tests
  let reactSwipeEl;

  const [realm, setRealm] = useState({
    label: '',
    value: '',
  });

  const handleRealmPicked = (e) => {
    console.log(e, 'realm');
    setRealm(realm);
  };

  return (
    <div className="main">

      <button type="button" onClick={() => reactSwipeEl.prev()}>Previous</button>
      <button type="button" onClick={() => reactSwipeEl.next()}>Next</button>

      <ReactSwipe className="carousel" swipeOptions={{ continuous: false }} ref={(el) => { reactSwipeEl = el; }}>
        <RealmSelection handleRealmPicked={handleRealmPicked} />
        <MainList />
        <Stats realm={realm} />
      </ReactSwipe>

    </div>
  );
}

export default AuctionPriceTrack;
