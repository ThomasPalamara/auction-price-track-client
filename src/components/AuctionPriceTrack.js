import React from 'react';
import RealmSelection from './RealmSelection';
import MainList from './MainList';
import DisplayStats from './DisplayStats';
import { Divider } from 'antd';

// TODO Ajouter la selection de language au store puis corrigÃ© celle en place et ajouter a display table

export default class AuctionPriceTrack extends React.Component {
  state = {
    realm : 'Ysondre',
    auctionHouse: null,
    loading: false,
  };
  handleRealmPicked = (realm) => {
    this.setState({ realm });
  }

  componentDidUpdate(prevProps, prevState) {
    if (this.state.realm && prevState.realm !== this.state.realm) {
      console.log('Fetching');
      this.setState({ loading: true });
      fetch(`/api/auctions/${this.state.realm}`)
        .then(res => res.json())
        .then(response => {
          this.setState({ loading: false });
          response.map((item) => {
            return item.auctions[0].prices.sort(function (a, b) { return (a.price > b.price) ? 1 : ((b.price > a.price) ? -1 : 0); });
          })
          return this.setState({ auctionHouse: response })
        }
      );
    }
  }

    render() {
    let displayPriceSection;
    if(!this.state.auctionHouse){
      displayPriceSection = <h3>Veuillez selectionner un serveur</h3>;
    }else{
      displayPriceSection = <DisplayStats auctionHouse={this.state.auctionHouse}/>
    }
    return (
      <div className="main">
        <RealmSelection loading={this.state.loading} handleRealmPicked={this.handleRealmPicked}/>
        <Divider/>
        <MainList/>
        <Divider/>
        {displayPriceSection}
      </div>
    );
  }
}
