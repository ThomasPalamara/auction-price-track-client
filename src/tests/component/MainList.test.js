import React from 'react';
import { shallow } from "enzyme";
import toJSON from 'enzyme-to-json';
import { MainList } from '../../components/MainList';
import recipes from '../fixtures/recipes';

test('Should render MainList component correctly', () => {
    const wrapper = shallow(<MainList/>);
    expect(toJSON(wrapper)).toMatchSnapshot();
})


//Penser a revoir le test avec les data filtrÃ©es

test('Should render MainList component correctly with recipes', () => {
    const wrapper = shallow(<MainList recipesList={recipes}/>);
    expect(toJSON(wrapper)).toMatchSnapshot();
    expect(wrapper.find('MainListItem').length).toBe(3);
})