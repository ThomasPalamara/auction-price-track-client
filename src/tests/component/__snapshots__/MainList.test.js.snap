// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render MainList component correctly 1`] = `
<div
  className="MainList"
>
  <h2
    className="step"
  >
    2/ Choose your Recipe
  </h2>
  <Connect(MainListFilter) />
  <Connect(ItemLanguage) />
  <Row
    className="container-item-display"
    gutter={0}
  >
    <Col
      span={24}
    >
      <Spin
        indicator={
          <Icon
            spin={true}
            style={
              Object {
                "fontSize": 24,
              }
            }
            type="loading"
          />
        }
        prefixCls="ant-spin"
        size="default"
        spinning={true}
        wrapperClassName=""
      >
        <ul
          className="list"
        >
          <FlipMovePropConverter
            delay={0}
            disableAllAnimations={false}
            duration={350}
            easing="ease-in-out"
            enterAnimation="elevator"
            getPosition={[Function]}
            leaveAnimation="elevator"
            maintainContainerHeight={false}
            staggerDelayBy={0}
            staggerDurationBy={0}
            typeName="div"
            verticalAlignment="top"
          />
        </ul>
      </Spin>
    </Col>
  </Row>
</div>
`;

exports[`Should render MainList component correctly with recipes 1`] = `
<div
  className="MainList"
>
  <h2
    className="step"
  >
    2/ Choose your Recipe
  </h2>
  <Connect(MainListFilter) />
  <Connect(ItemLanguage) />
  <Row
    className="container-item-display"
    gutter={0}
  >
    <Col
      span={24}
    >
      <Spin
        indicator={
          <Icon
            spin={true}
            style={
              Object {
                "fontSize": 24,
              }
            }
            type="loading"
          />
        }
        prefixCls="ant-spin"
        size="default"
        spinning={true}
        wrapperClassName=""
      >
        <ul
          className="list"
        >
          <FlipMovePropConverter
            delay={0}
            disableAllAnimations={false}
            duration={350}
            easing="ease-in-out"
            enterAnimation="elevator"
            getPosition={[Function]}
            leaveAnimation="elevator"
            maintainContainerHeight={false}
            staggerDelayBy={0}
            staggerDurationBy={0}
            typeName="div"
            verticalAlignment="top"
          >
            <MainListItem
              element={
                Object {
                  "isCustom": true,
                  "professions": Array [
                    "alchemy",
                    "inscription",
                  ],
                  "reagents": Array [
                    Object {
                      "blizzardId": 152510,
                      "quantity": 1,
                    },
                    Object {
                      "blizzardId": 152507,
                      "quantity": 1,
                    },
                    Object {
                      "blizzardId": 152511,
                      "quantity": 1,
                    },
                    Object {
                      "blizzardId": 152508,
                      "quantity": 1,
                    },
                    Object {
                      "blizzardId": 152505,
                      "quantity": 1,
                    },
                    Object {
                      "blizzardId": 152509,
                      "quantity": 1,
                    },
                    Object {
                      "blizzardId": 152506,
                      "quantity": 1,
                    },
                  ],
                }
              }
              key="0"
            />
            <MainListItem
              element={
                Object {
                  "craft": Object {
                    "blizzardId": 152641,
                    "quantity": 1,
                  },
                  "isCustom": false,
                  "professions": Array [
                    "alchemy",
                  ],
                  "reagents": Array [
                    Object {
                      "blizzardId": 152510,
                      "quantity": 5,
                    },
                    Object {
                      "blizzardId": 152507,
                      "quantity": 10,
                    },
                    Object {
                      "blizzardId": 152509,
                      "quantity": 15,
                    },
                  ],
                  "type": "flask",
                }
              }
              key="1"
            />
            <MainListItem
              element={
                Object {
                  "craft": Object {
                    "blizzardId": 153442,
                    "quantity": 1,
                  },
                  "isCustom": false,
                  "professions": Array [
                    "enchant",
                  ],
                  "reagents": Array [
                    Object {
                      "blizzardId": 152875,
                      "quantity": 15,
                    },
                    Object {
                      "blizzardId": 152876,
                      "quantity": 3,
                    },
                    Object {
                      "blizzardId": 152877,
                      "quantity": 4,
                    },
                  ],
                  "type": "ring",
                }
              }
              key="2"
            />
          </FlipMovePropConverter>
        </ul>
      </Spin>
    </Col>
  </Row>
</div>
`;
